pspipe_utils/covariance.py

```diff
@@ -313,6 +369,186 @@ def skew(cov, dir=1):
+        noise level in the target is estimated from the target values
+        themselves by the Gaussian process.
+    idx_arr : np.ndarray, optional
+        An array of size 0..n that indexes which data points are fit by the GP,
+        by default None. If None, all datapoints are used. Any unused datapoints
+        are not fit, but rather the target values are simply adopted at face 
+        value.
+    length_scale : float, optional
+        The initial RBF scale guess, by default 200.0.
+    length_scale_bounds : tuple, optional
+        The bounds of the RBF scale fit, by default (2, 2e4).
+    noise_level : float, optional
+        The initial white noise level guess, by default 0.01. Only used if 
+        var_diag is None.
+    noise_level_bounds : tuple, optional
+        The bounds of the white noise level guess, by default (1e-6, 1e1).
+        Only used if var_diag is None.
+    n_restarts_optimizer : int, optional
+        The number of optimizers to run to try to find a global minimum, 
+        by default 20.
+    random_state : int, RandomState instance, or None, optional
+        (From scikit-learn): determines random number generation used to
+        initialize the centers. Pass an int for reproducible results across
+        multiple function calls, by default 2.
+    remove_mean : bool, optional
+        If True, subtract the mean of the target prior to the fit and add it
+        back in the prediction, by default True.
+    return_gpr : bool, optional
+        If True, also return the sklearn.GaussianProcessRegressor object, by
+        default False.
+
+    Returns
+    -------
+    (n,) np.ndarray, {sklearn.GaussianProcessRegressor object}
+        The smoothed target, and optionally the object used to smooth.
+    """
+    if idx_arr is None:
+        idx_arr = np.arange(lb.size)
+
+    out = arr_diag.copy()
+
+    if len(idx_arr) > 0:
+        # fit only on the indexed data bins
+        X_train = lb[idx_arr, None]
+        y_train = arr_diag[idx_arr]
+
+        # remove mean
+        if remove_mean:
+            prior_mean = y_train.mean()
+            y_train = y_train - prior_mean
+        
+        # get the fit
+        if var_diag is None:
+            kernel = 1.0 * RBF(length_scale=length_scale, length_scale_bounds=length_scale_bounds) + \
+                WhiteKernel(noise_level=noise_level, noise_level_bounds=noise_level_bounds)
+            gpr = GaussianProcessRegressor(kernel=kernel, alpha=0,
+                                           n_restarts_optimizer=n_restarts_optimizer,
+                                           random_state=random_state)
+        else:
+            kernel = 1.0 * RBF(length_scale=length_scale, length_scale_bounds=length_scale_bounds)
+            gpr = GaussianProcessRegressor(kernel=kernel, alpha=var_diag[idx_arr],
+                                           n_restarts_optimizer=n_restarts_optimizer,
+                                           random_state=random_state)
```

Comment:
I may be wrong but to simplify a bit this part you can write something like
```python
kernel = 1.0 * RBF(length_scale=length_scale, length_scale_bounds=length_scale_bounds)
if var_diag is None:
    kernel += WhiteKernel(noise_level=noise_level, noise_level_bounds=noise_level_bounds)

alpha = var_diag[idx_arr] if var_diag is not None else 0
gpr = GaussianProcessRegressor(
    kernel=kernel,
    alpha=alpha,
    n_restarts_optimizer=n_restarts_optimizer,
    random_state=random_state,
)

```

Reply:
agreed